packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp

```diff
@@ -2185,11 +2080,91 @@ namespace Ifpack2 {
                                  pmv.extent(1), 
                                  pmv.extent(2), 
                                  vector_length/internal_vector_length),
-	vector_loop_size(vector_length/internal_vector_length)
+	vector_loop_size(vector_length/internal_vector_length),
+	Y_scalar_multivector(),
+	Z_scalar_multivector(),
+	df(damping_factor),
+	compute_diff(is_norm_manager_active)	
       {}
 
     public:
 
+      /// move packed multi vector into flat multi vector for computing residuals
+      KOKKOS_INLINE_FUNCTION
+      void
+      copyToFlatMultiVector(const member_type &member,
+			    const local_ordinal_type partidxbeg, // partidx for v = 0
+			    const local_ordinal_type npacks,
+			    const local_ordinal_type pri0,
+			    const local_ordinal_type v, // index with a loop of vector_loop_size
+			    const local_ordinal_type blocksize,
+			    const local_ordinal_type num_vectors) const {
+	const local_ordinal_type vbeg = v*internal_vector_length;
+	if (vbeg < npacks) {
+	  local_ordinal_type ri0_vals[internal_vector_length] = {};
+	  local_ordinal_type nrows_vals[internal_vector_length] = {};
+	  for (local_ordinal_type vv=vbeg,vi=0;vv<npacks && vi<internal_vector_length;++vv,++vi) {	
+	    const local_ordinal_type partidx = partidxbeg+vv;
+	    ri0_vals[vi] = partptr(partidx);
+	    nrows_vals[vi] = partptr(partidx+1) - ri0_vals[vi];
+	  }
+
+	  if (nrows_vals[0] == 1) {
+	    const local_ordinal_type j=0, pri=pri0;
+	    {
+	      for (local_ordinal_type vv=vbeg,vi=0;vv<npacks && vi<internal_vector_length;++vv,++vi) {
+		const local_ordinal_type ri0 = ri0_vals[vi];
+		const local_ordinal_type nrows = nrows_vals[vi];
+		if (j < nrows) {
+		  Kokkos::parallel_for
+		    (Kokkos::TeamThreadRange(member, blocksize), 
+		     [&](const local_ordinal_type &i) {
+		      const local_ordinal_type row = blocksize*lclrow(ri0+j)+i;
+		      for (local_ordinal_type col=0;col<num_vectors;++col) {
+			impl_scalar_type &y = Y_scalar_multivector(row,col);
+			const impl_scalar_type yd = X_internal_vector_values(pri, i, col, v)[vi] - y;
+			y  += df*yd;
+			
+			if (compute_diff) {
+			  impl_scalar_type &z = Z_scalar_multivector(row,col);
+			  const magnitude_type abs_yd = Kokkos::ArithTraits<impl_scalar_type>::abs(yd); 
+			  z = abs_yd*abs_yd;
+			}
```

Comment:
@mhoemmen The norm computation is right. here I compute square of diff value.

Reply:
ok